pkg/reconciler/trustedkeys/trustedkeys.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trustedkeys
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/policy-controller/pkg/apis/config"
+	"github.com/sigstore/policy-controller/pkg/apis/policy/v1alpha1"
+	trustedkeysreconciler "github.com/sigstore/policy-controller/pkg/client/injection/reconciler/policy/v1alpha1/trustedkeys"
+	"github.com/sigstore/policy-controller/pkg/reconciler/trustedkeys/resources"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+)
+
+// Reconciler implements ConfigMap reconciler.
+// TUFRoot resources.
+type Reconciler struct {
+	configmaplister corev1listers.ConfigMapLister
+	kubeclient      kubernetes.Interface
+}
+
+// Check that our Reconciler implements Interface as well as finalizer
+var _ trustedkeysreconciler.Interface = (*Reconciler)(nil)
+var _ trustedkeysreconciler.Finalizer = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, trustedKeys *v1alpha1.TrustedKeys) reconciler.Event {
+
+	var sigstoreKeys *config.SigstoreKeys
+	var err error
+	if trustedKeys.Spec.Repository != nil {
+		sigstoreKeys, err = r.getSigstoreKeysFromMirrorFS(ctx, trustedKeys.Spec.Repository)
+	} else if trustedKeys.Spec.Remote != nil {
+		sigstoreKeys, err = r.getSigstoreKeysFromRemote(ctx, trustedKeys.Spec.Remote)
+	} else if trustedKeys.Spec.SigstoreKeys != nil {
+		sigstoreKeys = &config.SigstoreKeys{}
+		sigstoreKeys.ConvertFrom(ctx, trustedKeys.Spec.SigstoreKeys)
+	} else {
+		err = fmt.Errorf("Invalid TrustedKeys entry: %s missing repository,remote, and sigstoreKeys", trustedKeys.Name)
+		logging.FromContext(ctx).Errorf("Invalid TrustedKeys entry: %s missing repository,remote, and sigstoreKeys", trustedKeys.Name)
+	}
```

Comment:
I assume validation checks that these are mutually exclusive (oneof)?

Reply:
Yep!